### PR TITLE
fix: removal of account permissions when removing account

### LINKED ISSUE #21099: [Sentry] Error: eth_accounts error: Received unrecognized address: "0x**".
Sentry Issue: [METAMASK-X75Z](https://metamask.sentry.io/issues/4505615114/?referrer=github_integration)

```
Error: eth_accounts error: Received unrecognized address: "0x**".
  at Error (/metamask/runtime-lavamoat.js:7838:17)
  at forEach (app/scripts/controllers/permissions/specifications.js:203:13)
  at validateCaveatAccounts (app/scripts/controllers/permissions/specifications.js:195:12)
  at <object>.validator (app/scripts/controllers/permissions/specifications.js:69:9)
  at <object>.addPermittedAccount (app/scripts/controllers/permissions/background-api.js:20:28)
...
(25 additional frame(s) were not displayed)
```

Comment: This might have been introduced by this [commit](https://github.com/MetaMask/metamask-extension/commit/9487e82e93edad4d7bf59eb7aae46c6bf3be4032)

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This PR addresses the issue where account caveats remain after an account has been removed. During the account removal process, the permission is not being correctly removed because the targetAddress is a checksummed address and is being compared to a list of lowercase addresses. Consequently, the address is not found and thus not removed. When attempting to connect to a dApp with a dangling address, this address is never found, blocking any account from connecting to the site again.

## **Related issues**

Fixes: [21099](https://github.com/MetaMask/metamask-extension/issues/21099)
Resolves: [293](https://github.com/MetaMask/accounts-planning/issues/293)

## **Manual testing steps**

1. Import SRP
2. Add Hardware wallet (Ledger tested but could not reproduce with a Snap account from SSK v1.1.1)
3. Connect to [test dapp ](https://metamask.github.io/test-dapp/) and open console
4. Conduct a V4 typed data signature and verify recovery result
5. Return to add hardware flow and forget device
6. Go back to test dapp from full screen add hardware
7. Then open pop-up again by tapping on fox
8. Tap `Connect` to connect Account 1 or whatever account is now active

### **Before**

https://github.com/MetaMask/metamask-extension/assets/96463427/ed9ef1e6-ed93-4899-a108-8f5ec5cbd17f

### **After**

https://github.com/MetaMask/metamask-extension/assets/96463427/d13a43ee-63fb-4919-bff1-4862594a17bc

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by ccharly** [comment]
FYI @montelaidev, I've fixed a checkbox in the main description to avoid having the "INVALID-TEMPLATE-PR" label.

**Review by ccharly — DISMISSED** [review]
Manually tested, on `develop` first to check the bug and then with the fix. With this PR I am able to re-connect any account using the test dapp (meaning there's no more dangling accounts remaining after the removal of the HW wallet account)!

Was a tricky one :D gj!

**Review by ccharly — DISMISSED** [review]
Haven't manually re-tested, but "normalizing" both address to the same form should not change the behavior (assuming `toChecksumAddress` works as intended 🙈)

**Review by owencraston — DISMISSED** [review]

**Review by gantunesr — COMMENTED** [review]
The code LGTM but I was not able to reproduce the bug in my end with a QR wallet.

@montelaidev can you also add a recording of the bug and the fix for documentation purposes in the PR description?

**Comment by montelaidev** [comment]
> The code LGTM but I was not able to reproduce the bug in my end with a QR wallet.
> 
> @montelaidev can you also add a recording of the bug and the fix for documentation purposes in the PR description?

Just added the before and after videos

**Review by ccharly — DISMISSED** [review]
LGTM, haven't re-tested manually, but the video is exactly what I had (prior to the change of using `toChecksumAddress`, which should not have any impact on the behavior of this fix)

**Comment by ccharly** [comment]
@montelaidev I've fixed the PR template issue (was missing the section for screenshots/recordings), so the CI should be happy now! :)

File: app/scripts/controllers/permissions/caveat-mutators.js [hunk 1]
```diff
@@ -24,7 +25,8 @@ export const CaveatMutatorFactories = {
  */
 function removeAccount(targetAccount, existingAccounts) {
   const newAccounts = existingAccounts.filter(
-    (address) => address !== targetAccount,
+    (address) =>
+      toChecksumAddress(address) !== toChecksumAddress(targetAccount),
```
Comment (FrederikBolding) [inline app/scripts/controllers/permissions/caveat-mutators.js:29]: Could we compute the checksum of the `targetAccount` outside the loop?

Comment (montelaidev) [reply app/scripts/controllers/permissions/caveat-mutators.js]: Thanks, thats a good catch!

**Review by ccharly — APPROVED** [review]

**Review by gantunesr — APPROVED** [review]

**Pull request merged**